src/raw/mod.rs

```diff
@@ -1206,6 +1222,90 @@ impl<A: Allocator + Clone> RawTableInner<A> {
+        group: &Group,
+        probe_seq: &ProbeSeq,
+    ) -> Option<usize> {
+        let bit = group.match_empty_or_deleted().lowest_set_bit();
+
+        if likely(bit.is_some()) {
+            let mut index = (probe_seq.pos + bit.unwrap()) & self.bucket_mask;
+
+            // In tables smaller than the group width, trailing control
+            // bytes outside the range of the table are filled with
+            // EMPTY entries. These will unfortunately trigger a
+            // match, but once masked may point to a full bucket that
+            // is already occupied. We detect this situation here and
+            // perform a second scan starting at the beginning of the
+            // table. This second scan is guaranteed to find an empty
+            // slot (due to the load factor) before hitting the trailing
+            // control bytes (containing EMPTY).
+            if unlikely(is_full(*self.ctrl(index))) {
+                debug_assert!(self.bucket_mask < Group::WIDTH);
+                debug_assert_ne!(probe_seq.pos, 0);
+                index = Group::load_aligned(self.ctrl(0))
+                    .match_empty_or_deleted()
+                    .lowest_set_bit_nonzero();
+            }
+
+            Some(index)
+        } else {
+            None
+        }
+    }
+
+    /// Searches for an element in the table,
+    /// or a potential slot where that element could be inserted.
+    #[inline]
+    pub fn find_potential_inner(
+        &self,
+        hash: u64,
+        eq: &mut dyn FnMut(usize) -> bool,
+    ) -> (usize, bool) {
+        unsafe {
+            let mut tombstone = None;
+
+            let h2_hash = h2(hash);
+            let mut probe_seq = self.probe_seq(hash);
+
+            loop {
+                let group = Group::load(self.ctrl(probe_seq.pos));
+
+                for bit in group.match_byte(h2_hash).into_iter() {
+                    let index = (probe_seq.pos + bit) & self.bucket_mask;
+
+                    if likely(eq(index)) {
+                        return (index, true);
+                    }
+                }
+
+                let index = self.find_insert_slot_in_group(&group, &probe_seq);
+
+                if likely(index.is_some()) {
+                    // Only stop the search if the group is empty. The element might be
+                    // in a following group.
+                    if likely(group.match_empty().any_bit_set()) {
+                        // Use a tombstone if we found one
+                        return if unlikely(tombstone.is_some()) {
+                            (tombstone.unwrap(), false)
+                        } else {
+                            (index.unwrap(), false)
+                        };
+                    } else {
+                        // We found a tombstone, record it so we can return it as a potential
+                        // insertion location.
+                        tombstone = index;
+                    }
+                }
```

Comment:
I would reorganize this code like this, which makes it much clearer:
```rust
// We didn't find the element we were looking for in the group, try to get an
// insertion slot from the group if we don't have one yet.
if insert_slot.is_none() {
    insert_slot = self.find_insert_slot_in_group(&group, &probe_seq);
}

// Only stop the search if the group contains at least one empty element.
// Otherwise, the element that we are looking for might be in a following group.
if likely(group.match_empty().any_bit_set()) {
    return (insert_slot.unchecked_unwrap(), false);
}
```
(`tombstone` is renamed to `insert_slot`)